cmd/release/main.go

```diff
@@ -26,123 +23,170 @@ import (
+		dryRun:             *dryRun,
+		rootDir:            root,
+	}
+	if err := cmd.run(); err != nil {
 		log.Fatalln(err.Error())
 	}
 }
 
-func run(root string, minisignPrivateKey string, dryRun bool) error {
+type command struct {
+	minisignPrivateKey string
+	minisignPublicKey  string
+	githubReleaseOwner release.GithubOwner
+	dryRun             bool
+	rootDir            string
+}
+
+func (c *command) run() error {
 	// Create temporary directory
 	tmpDir, err := os.MkdirTemp("", "plugins-release")
 	if err != nil {
 		return fmt.Errorf("failed to create temporary directory: %w", err)
 	}
 	log.Printf("created tmp dir: %s", tmpDir)
-	if !dryRun {
-		defer func() {
-			if err := os.RemoveAll(tmpDir); err != nil {
-				log.Printf("failed to remove %q: %v", tmpDir, err)
-			}
-		}()
-	}
+	defer func() {
+		if c.dryRun {
+			return
+		}
+		if err := os.RemoveAll(tmpDir); err != nil {
+			log.Printf("failed to remove %q: %v", tmpDir, err)
+		}
+	}()
 
 	ctx := context.Background()
-	client := newGitHubClient(ctx)
-	latestRelease, err := getLatestRelease(ctx, client)
-	if err != nil {
+	client := release.NewClient(ctx)
+	latestRelease, err := client.GetLatestRelease(ctx, c.githubReleaseOwner, release.GithubRepoPlugins)
+	if err != nil && !errors.Is(err, release.ErrNotFound) {
 		return fmt.Errorf("failed to retrieve latest release: %w", err)
 	}
 
-	releases, err := loadPluginReleases(ctx, client, latestRelease)
+	var privateKey minisign.PrivateKey
+	if minisignPrivateKeyPassword := os.Getenv("MINISIGN_PRIVATE_KEY_PASSWORD"); minisignPrivateKeyPassword != "" {
+		privateKey, err = minisign.PrivateKeyFromFile(minisignPrivateKeyPassword, c.minisignPrivateKey)
+		if err != nil {
+			return err
+		}
+	}
+	publicKey, err := c.loadMinisignPublicKeyFromFileOrPrivateKey(privateKey)
 	if err != nil {
+		return err
+	}
+	releases, err := client.LoadPluginReleases(ctx, latestRelease, publicKey)
+	if err != nil && !errors.Is(err, release.ErrNotFound) {
 		return fmt.Errorf("failed to determine latest plugin releases: %w", err)
 	}
 	if releases == nil {
 		log.Printf("no current release found")
-		releases = &PluginReleases{}
-	}
-
-	pluginNameVersionToRelease := make(map[pluginNameVersion]PluginRelease, len(releases.Releases))
-	for _, release := range releases.Releases {
-		key := pluginNameVersion{name: release.PluginName, version: release.PluginVersion}
-		if _, ok := pluginNameVersionToRelease[key]; ok {
-			return fmt.Errorf("duplicate plugin discovered in releases file: %+v", key)
-		}
-		pluginNameVersionToRelease[key] = release
+		releases = &release.PluginReleases{}
 	}
 
-	var newPlugins []PluginRelease
-	var existingPlugins []PluginRelease
 	now := time.Now().UTC().Truncate(time.Second)
 	releaseName, err := calculateNextRelease(now, latestRelease)
 	if err != nil {
 		return fmt.Errorf("failed to determine next release name: %w", err)
 	}
 
-	if err := plugin.Walk(root, func(plugin *plugin.Plugin) error {
```

Comment:
Kept pushing up against the edge of cyclomatic complexity linter so moved the bulk of the run method into `calculateNewReleasePlugins`.